nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/PutMongo.java

```diff
@@ -196,6 +237,33 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
         }
     }
 
+    private void removeUpdateKeys(String updateKeyParam, Map doc) {
+        String[] parts = updateKeyParam.split(",[\\s]*");
+        for (String part : parts) {
+            if (part.contains(".")) {
```

Comment:
Forgot to raise this earlier. Why are we doing a check if it contains a `.` ? It will work only for complex keys and simple key will be ignored i.e. not removed, right?

Reply:
For this input:

```
{
    "name": "John Smith",
    "department": "Engineering"
}
```

It makes no sense to remove `name` if we're doing a full document update using the `name key.

Now consider this complex document:

```
{
    "name": "John Smith",
    "department": "Engineering",
    "contacts": {
         "email": "john.smith@test.com"
     }
}
```

To search on `email`, we have to submit this payload with the lookup key being `contacts.email`:

```
{
    "contacts.email": "john.smith@test.com",
    "name": "John Smith",
    "department": "Engineering",
    "contacts": {
         "email": "john.smith@test.com"
     }
}
```

Mongo cannot do a lookup using this: `{ "contacts": { "email": "john.smith@test.com" }}`

So if we don't remove the complex lookup key, we are leaving extraneous information in the document that almost certainly has no value to the user.

Now maybe we'll get an angry ticket complaining that they can't do periods in the key names, but I've never seen normal use cases where developers do that. The whole idea of creating complex key names for real data using periods and such flies in the face of how JSON is supposed to work.

Reply:
No.. what I meant was, regardless of whether we have a complex key or simple key, we don't need them to exist in the parsed filter query, right? So why don't we simply remove this `if (part.contains("."))` check, basically meaning, remove the key from the `doc`. No?

Reply:
> we don't need them to exist in the parsed filter query, right?

We do when replacing one document with another. Say you have a user account document, and the update key is `username`. If the user is doing full document replacement, and we remove the update key, the `username` field is going to be removed from the document that is used in the substitution. So then the poor end users will be searching through an accounts collection looking for documents with no usernames.

Reply:
Also, again, complex key names simply wouldn't have any value to over 99% of Mongo users. As I said, I'm sure there's a few random users that want to chuck something like this in Mongo:

```
{
    "user.name": "john.smith",
    "user.contact.email": "john.smith@test.com"
}
```

Accommodating that would be detrimental to providing a consistent tool for the vast majority of Mongo users.

Reply:
Understood. So this is critical when doing replace i.e. the mode is set to `UPDATE_WITH_DOC`. Correct?

Reply:
Yes. The only time it wouldn't apply is if a user wanted to remove a field like they misnamed the `username` field as `yusername`. They can fix that with:

```
{
    "yusername": "....",
    "$set": {
         "username": ".....",
     },
     "$unset": {
         "yusername": 1
     }
}
```

Reply:
TBH, the sanest way to handle such mistakes would be to stop ingestion and run some JavaScript code to update the database properly.